### DISCUSSION TITLE
Help with protobuf decoding by schema registry

### DISCUSSION THREAD

**Discussion by mihelanjelo** [provectus/kafka-ui#1680] [Q&A] [answered]
Hello, i have problem with setting up proto schemas for topic.
I have connected schema registry confluentinc/cp-schema-registry:7.0.1
I set schema for key like topic_name-key and for value like topic_name-value
, but still messages are not shown decoded.
I checked schemas many times and decoded messages in code by google lib successfully.
Sometimes i saw error in log like:
```
org.apache.kafka.common.errors.SerializationException: Error deserializing Protobuf message for id -1
Caused by: org.apache.kafka.common.errors.SerializationException: Unknown magic byte!
```
Please, explain by steps how to set it up properly

**Comment by Haarolean** [comment]
Hi, could you please share an example schema and a message to reproduce this? Thanks.

**Reply by mihelanjelo** [reply]
ok, but before i have question
why do messages look like that?
on UI
<img width="1168" alt="Screenshot 2022-02-25 at 23 20 02" src="https://user-images.githubusercontent.com/20622825/155796166-2b668739-44ca-442d-82f3-9641f9e125cb.png">
in file
<img width="680" alt="Screenshot 2022-02-25 at 23 20 35" src="https://user-images.githubusercontent.com/20622825/155796255-8649b596-7acf-4e96-8096-2b68a2266a17.png">
when i read protobuf message from topic in code by python lib for kafka it looks like 
`b'\x08\xee\xd5\xe5\x8a\x01\x12\x0c\x08\xff\xe5\xdd\x90\x06\x10\xc4\xd9\xb3\xb2\x01Z*\n(\n$13696e0e-4c75-49aa-9c83-0a6faf5074ee\x10\x01r\x0c0_7311196171'`

**Reply by Haarolean** [reply]
We need a schema and a message example to reproduce this.

**Reply by mihelanjelo** [reply]
Schema:
  Key:  
    syntax = "proto3";
    message EventKey {
      int64 some_id = 1;
    }  
  Value :
    syntax = "proto3";
    message EventValue {
      int64 some_id = 1;
      Timestamp date = 2;
      string some_parent_id = 14;
    }
    message Timestamp {
      int64 seconds = 1;
      int32 nanos = 2;
    }

Message:
  Encoded:
    Key: b'\x08\xb9\xdc\xc1\x7f'
    Value: b'\x08\xee\xd5\xe5\x8a\x01\x12\x0c\x08\xff\xe5\xdd\x90\x06\x10\xc4\xd9\xb3\xb2\x01Z*\n(\n$13696e0e-4c75-49aa-9c83-0a6faf5074ee\x10\x01r\x0c0_7311196171'
  
  Decoded:
    Key: 
     some_id: 267415097
    
    Value: 
     some_id: 291072750
    date {
      seconds: 1645703935
      nanos: 374140100
    }
    some_parent_id: "0_7311196171"

**Reply by mihelanjelo** [reply]
I tried to produce massage in kafka-ui and decoding worked, but i need to read decoded messages that are already in topic and encoded in protobuf

**Reply by Haarolean** [reply]
Thanks for the info, we'll take a look. I'll convert this into an issue.

**Reply by mihelanjelo** [reply]
Thank you!

**Comment by iliax** [comment]
Hello @mihelanjelo!
problem is likely in serializer you use when writing to kafka.
In raw bytes message should look like [zero byte, 4 bytes - schema id, remaining payload bytes] see https://docs.confluent.io/platform/current/schema-registry/serdes-develop/index.html#wire-format. 

Usually kafka libraries hide first 5 bytes generating for end user. In your case you likely skip 5-bytes prefix, so kafka deserializer can't detect schema for parsing. In such scenario kafka-ui falls back to string deserializing.

See python AvroSerializer usage here: https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/avro_producer.py

**Reply by mihelanjelo** [reply]
Hello. Why avro serializer? I need protobuf serializer and i set schema for protobuf in schema registry.

**Reply by mihelanjelo** [reply]
i can't understand why i can read and decode protobuf messages in my python code (kafka-python + protobuf libs) but in Kafka-UI it looks like encoded differently and unable to decode (

**Comment by iliax** [comment]
@mihelanjelo You can read and decode protobuf messages in your code because you know schema of you messages, but kafka-ui does not. To solve such cases schema registry exists.  

Schema registry is needed to mark every message you write to kafka with schema id used for serialization. Later when reading messages (by any other consumer) corresponding schema will be extracted from registry and used for deserialzation. Please see concepts described here -> https://docs.confluent.io/platform/current/schema-registry/index.html. 

Sorry for pointing you to wrong producing-> here is an example of protobuf producer schemaregistry usage https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/protobuf_producer.py#L76. 

From my understanding currently you just serialize  your message  using protobuf and write to kafka as raw bytes - in that case consumer have no ability to get schema, even it is created in schema-registry.

---

If you want to write raw  bytes messages, serialized via protobuf (without using schemaregistry) you can pass kafka-ui topic->protofile mapping. This mapping will be used for messages deserialization by kafka-ui and you will be able to see messages content in ui. (this is not documented, and only values deserialisation supported for such topics). Let me know I will write how to pass such config.

**Reply by mihelanjelo** [reply]
Nooo, i'm using schema registry, set protobuf schemas with names like topic-value and topic-key, when i try to produce messages in kafka-ui i see generated templates and it means for me that schemas set. But i still can not see readable messages (

**Reply by iliax** [reply]
kafka-ui rendering templates based on schemas found in schemaregistry, but it doesnt affecdt messages that you already written to kafka.
1. can you try to submit message via kafka-ui ? It will be written in right format and you will see new message in ui.
2. what library you use for producing messages to kafka? maybe you can share a code snippet

**Reply by mihelanjelo** [reply]
i can not share code that produce messages, i do not even have permissions to see him)) i am trying to reproduce decoding in my own python code and it works, i see that in kafka-ui messages encoded differently as i said and can not understand why. Maybe it is possible to handle that behavior in kafka-ui? I can't affect existing message sending (

**Reply by mihelanjelo** [reply]
python code example that can decode this message:
```
from protofuzz import pbimport

key_schema = """syntax = "proto3";
message EventKey {
  int64 some_id = 1;
}
"""
value_schema = """syntax = "proto3";
message EventValue {
  int64 some_id = 1;
  Timestamp date = 2;
  string some_parent_id = 14;
}
message Timestamp {
  int64 seconds = 1;
  int32 nanos = 2;
}
"""


if __name__ == '__main__':
    key = b'\x08\xb9\xdc\xc1\x7f'
    value = b'\x08\xee\xd5\xe5\x8a\x01\x12\x0c\x08\xff\xe5\xdd\x90\x06\x10\xc4\xd9\xb3\xb2\x01Z*\n(\n$13696e0e-4c75-49aa-9c83-0a6faf5074ee\x10\x01r\x0c0_7311196171'
    key_pb = pbimport.from_string(key_schema).EventKey()
    value_pb = pbimport.from_string(value_schema).EventValue()
    decoded_value = value_pb.FromString(value)
    decoded_key = key_pb.FromString(key)
    print('Key: \n', decoded_key)
    print('Value: \n', decoded_value)

```
Output :
Key: 
 some_id: 267415097

Value: 
 some_id: 291072750
date {
  seconds: 1645703935
  nanos: 374140100
}
some_parent_id: "0_7311196171"

**Reply by iliax** [reply]
@mihelanjelo from payload you provided - you currently not using schemaregistry when writing and reading to kafka. When using SR each k&v will start with [zero byte, 4 bytes (schema id)] prefix (see links above). So, currently your messages are not connected with SR at all. Because of this kafka ui can't deserialize messages in proper way.

**Reply by mihelanjelo** [reply]
@iliax so all producer must be connected to SR?

**Reply by iliax** [reply]
yes, they should register schemas in SR before writing to kafka like [here](https://github.com/confluentinc/confluent-kafka-python/blob/master/examples/protobuf_producer.py#L76)

**Reply by mihelanjelo** [reply]
seems like schemas ids are not so necessary if kafka-ui links schemas by topic name

**Reply by iliax** [reply]
yes, theoretically kafka-ui can get schema by topic name, but it is wrong usage of SR. If for example you will produce message from kafka-ui, message will be generated in proper way (and will be visible), but it likely will break consumers.

**Reply by mihelanjelo** [reply]
so now all producers and consumers must be connected to SR. But seems possible to create an option to work without those ids in messages as kafka-ui for itself can get ids by topic name...

**Reply by iliax** [reply]
I think it can lead to SR misusage. We will discuss it think how it can be implemented. Thank you for suggestion.

**Reply by messart** [reply]
Hi @iliax,

You wrote before 
"If you want to write raw bytes messages, serialized via protobuf (without using schemaregistry) you can pass kafka-ui topic->protofile mapping. This mapping will be used for messages deserialization by kafka-ui and you will be able to see messages content in ui. (this is not documented, and only values deserialisation supported for such topics). Let me know I will write how to pass such config."

Could you please describe the way how to apply this mapping? As we have the case that you've described with the topic that contains raw serialized via protobuf messages (without using schemaregistry) and we'd like to view them in kafka-ui in the deserialized way.

Thanks in advance.

**Reply by iliax** [reply]
Hi @messart, here are things you need to do:
1. mount proto file with messages that are used to kafka-ui docker container  (you can paste all messages in one file)
2. set properties for mapping, example: 
```
KAFKA_CLUSTERS_0_PROTOBUF_FILE=/path/to/mounted/protofile.proto
// this proto message will be used for values parsing if there is no specific topic mapping (default)
KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME=DefaultMessageName
// this proto message will be used for key parsing if there is no specific topic mapping (default)
KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME_FOR_KEY=DefaultMessageNameForKey

//you can also specify proto message name for specific topic's keys&values:
KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME_BY_TOPIC_TOPICNAME1=Topic1MessageName
KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME_FOR_KEY_BY_TOPIC_TOPICNAME1=Topic1KeyMessageName

KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME_BY_TOPIC_ANOTERTOPICNAME=AonotherTopicMessageName
KAFKA_CLUSTERS_0_PROTOBUF_MESSAGE_NAME_FOR_KEY_BY_TOPIC_ANOTERTOPICNAME=AonotherTopicKeyMessageName
```

**Comment by Haarolean** [comment, marked as answer]
Latest [docs](https://docs.kafka-ui.provectus.io/configuration/serialization-serde) on serialization
